src/singledocparser.cpp

```diff
@@ -46,6 +46,9 @@ void SingleDocParser::HandleDocument(EventHandler& eventHandler) {
 }
 
 void SingleDocParser::HandleNode(EventHandler& eventHandler) {
+  if (depth > depth_limit) {
+    throw ParserException(m_scanner.mark(), ErrorMsg::BAD_FILE);
```

Comment:
Note that throwing will eliminate the possibility of re-using this object because it does not reset depth.

I suggest using a depth guard type of class and instantiate that at the entrypoint of the function. A quick example:

```
// suggested type, adapt as necessary
class deep_recursion : public std::runtime_error {
  int at_depth_;
public:
  using std::runtime_error::runtime_error;
  deep_recursion(int at_depth) : std::runtime_error{std::string{"recursion reached depth "} + std::to_string(at_depth)}, at_depth_(at_depth) {}
  int at_depth() const { return at_depth_; }
  // might need virtual dtor if change inheritance tree
}

// also a suggested type, adapt as necessary
template <int max_depth = 2000>
class depth_guard {
  int & depth_;
public:
  using deep_recursion_exception = deep_recursion; // or whatever type
  depth_guard(int & depth) : depth_(depth) {
    ++depth_;
    if ( max_depth <= depth_ ) {
        throw deep_recursion_exception{depth}; // or whatever constructor
    }
  }
  ~depth_guard() {
    --depth_;
  }
  int current_depth() const { return depth_; }
}

// example use:
void SingleDocParser::HandleNode(EventHandler& eventHandler) {
  using DepthGuard = depth_guard<2000>;
  DepthGuard guard{depth};
  ...
  // code here is guaranteed to not be too 'deep'
```

That also means that you don't need to explicitly increment and decrement depth everywhere you're calling the function. That will keep maintenance cost low and reduce human error opportunities. It can also be reused elsewhere ;)

Honestly though I think a better solution is to not use recursion: stack depth is not consistent across platforms. I've personally encountered stack size problems when porting stuff from Linux to OS X.